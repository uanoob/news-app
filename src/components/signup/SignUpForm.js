import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import {
  checkValidityLength,
  checkValidityEmail,
} from '../../utils/validator.utils';
import { signUp } from '../../store/actions';

const styles = theme => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
    margin: 'auto',
    maxWidth: 500,
  },
  textField: {
    marginLeft: theme.spacing.unit,
    marginRight: theme.spacing.unit,
  },
});

class SignUpForm extends React.Component {
  state = {
    name: '',
    nameTouched: false,
    nameIsValid: false,
    email: '',
    emailTouched: false,
    emailIsValid: false,
    password: '',
    passwordTouched: false,
    passwordIsValid: false,
  };

  handleChange = name => (event) => {
    const touched = `${name}Touched`;
    const isValid = `${name}IsValid`;
    const checkValidity = () => {
      if (name === 'email') {
        return checkValidityEmail(event.target.value);
      }
      return checkValidityLength(name, event.target.value);
    };
    this.setState({
      [name]: event.target.value,
      [touched]: true,
      [isValid]: checkValidity(),
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const { name, email, password } = this.state;
    const { onSignUp } = this.props;
    onSignUp(name, email, password);
  };

  render() {
    const { classes } = this.props;
    const {
      name,
      nameTouched,
      nameIsValid,
      email,
      emailTouched,
      emailIsValid,
      password,
      passwordTouched,
      passwordIsValid,
    } = this.state;
    return (
      <form
        className={classes.container}
        noValidate
        autoComplete="off"
        onSubmit={this.handleSubmit}
      >
        <TextField
          id="outlined-name-input"
          label="Name"
          className={classes.textField}
          value={name}
          onChange={this.handleChange('name')}
          type="text"
          autoComplete="current-name"
          margin="normal"
          variant="outlined"
          error={!nameIsValid && nameTouched}
        />
        <TextField
          id="outlined-email-input"
          label="Email"
          className={classes.textField}
          value={email}
          onChange={this.handleChange('email')}
          type="email"
          name="email"
          autoComplete="email"
          margin="normal"
          variant="outlined"
          error={!emailIsValid && emailTouched}
        />
        <TextField
          id="outlined-password-input"
          label="Password"
          className={classes.textField}
          value={password}
          onChange={this.handleChange('password')}
          type="password"
          autoComplete="current-password"
          margin="normal"
          variant="outlined"
          error={!passwordIsValid && passwordTouched}
        />
        <Button
          type="submit"
          variant="contained"
          color="primary"
          className={classes.button}
          disabled={!(emailIsValid && passwordIsValid)}
        >
          Sign Up
        </Button>
      </form>
    );
  }
}

SignUpForm.propTypes = {
  classes: PropTypes.shape({
    container: PropTypes.string.isRequired,
    textField: PropTypes.string.isRequired,
  }).isRequired,
  onSignUp: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  onSignUp: signUp,
};

export default connect(
  null,
  mapDispatchToProps,
)(withStyles(styles)(SignUpForm));
