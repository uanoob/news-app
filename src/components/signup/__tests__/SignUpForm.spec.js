import React from 'react';
import { shallow } from 'enzyme';
import { SignUpFormComponent } from '../SignUpForm';

describe('SignUp', () => {
  const mockSignUp = jest.fn();
  const initialState = {
    name: '',
    nameTouched: false,
    nameIsValid: false,
    email: '',
    emailTouched: false,
    emailIsValid: false,
    password: '',
    passwordTouched: false,
    passwordIsValid: false,
  };
  const props = {
    classes: {
      container: '',
      textField: '',
    },
    onSignUp: mockSignUp,
  };
  const wrapper = shallow(<SignUpFormComponent {...props} />);

  it('render correctly SignUpFormComponent component', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('initialize SignUpFormComponent with initial state', () => {
    expect(wrapper.state()).toEqual(initialState);
  });
  describe('valid name input', () => {
    const name = 'doe';
    const event = {
      target: { name: 'name', value: name },
      preventDefault: () => {},
    };
    beforeEach(() => {
      wrapper.find('#outlined-name-input').simulate('change', event);
    });

    it('updates name field in state', () => {
      expect(wrapper.state('name')).toEqual(name);
    });
    it('updates nameTouched field in state', () => {
      expect(wrapper.state('nameTouched')).toEqual(true);
    });
    it('updates nameIsValid field in state', () => {
      expect(wrapper.state('nameIsValid')).toEqual(true);
    });
  });
  describe('invalid name input', () => {
    const name = 'do';
    const event = {
      target: { name: 'name', value: name },
      preventDefault: () => {},
    };
    beforeEach(() => {
      wrapper.find('#outlined-name-input').simulate('change', event);
    });

    it('updates name field in state', () => {
      expect(wrapper.state('name')).toEqual(name);
    });
    it('updates nameTouched field in state', () => {
      expect(wrapper.state('nameTouched')).toEqual(true);
    });
    it('updates nameIsValid field in state', () => {
      expect(wrapper.state('nameIsValid')).toEqual(false);
    });
  });
  describe('valid email input', () => {
    const email = 'doe@icloud.com';
    const event = {
      target: { name: 'email', value: email },
      preventDefault: () => {},
    };
    beforeEach(() => {
      wrapper.find('#outlined-email-input').simulate('change', event);
    });

    it('updates email field in state', () => {
      expect(wrapper.state('email')).toEqual(email);
    });
    it('updates emailTouched field in state', () => {
      expect(wrapper.state('emailTouched')).toEqual(true);
    });
    it('updates emailIsValid field in state', () => {
      expect(wrapper.state('emailIsValid')).toEqual(true);
    });
  });
  describe('invalid email input', () => {
    const email = 'doe@icloud';
    const event = {
      target: { name: 'email', value: email },
      preventDefault: () => {},
    };
    beforeEach(() => {
      wrapper.find('#outlined-email-input').simulate('change', event);
    });

    it('updates email field in state', () => {
      expect(wrapper.state('email')).toEqual(email);
    });
    it('updates emailTouched field in state', () => {
      expect(wrapper.state('emailTouched')).toEqual(true);
    });
    it('updates emailIsValid field in state', () => {
      expect(wrapper.state('emailIsValid')).toEqual(false);
    });
  });
  describe('valid password input', () => {
    const password = 'password';
    const event = {
      target: { name: 'password', value: password },
      preventDefault: () => {},
    };
    beforeEach(() => {
      wrapper.find('#outlined-password-input').simulate('change', event);
    });

    it('updates password field in state', () => {
      expect(wrapper.state('password')).toEqual(password);
    });
    it('updates passwordTouched field in state', () => {
      expect(wrapper.state('passwordTouched')).toEqual(true);
    });
    it('updates passwordIsValid field in state', () => {
      expect(wrapper.state('passwordIsValid')).toEqual(true);
    });
  });
  describe('invalid password input', () => {
    const password = 'pa';
    const event = {
      target: { name: 'password', value: password },
      preventDefault: () => {},
    };
    beforeEach(() => {
      wrapper.find('#outlined-password-input').simulate('change', event);
    });

    it('updates password field in state', () => {
      expect(wrapper.state('password')).toEqual(password);
    });
    it('updates passwordTouched field in state', () => {
      expect(wrapper.state('passwordTouched')).toEqual(true);
    });
    it('updates passwordIsValid field in state', () => {
      expect(wrapper.state('passwordIsValid')).toEqual(false);
    });
  });
  describe('when submiting the form', () => {
    const event = { preventDefault: () => {} };
    beforeEach(() => {
      wrapper.find('form').simulate('submit', event);
    });
    it('calls the onSignUp', () => {
      expect(mockSignUp).toBeCalledTimes(1);
    });
  });
  describe('when clicking the Sign Up button', () => {
    const event = { preventDefault: () => {} };
    beforeEach(() => {
      wrapper.find('#signup-button').simulate('click', event);
    });
    it('calls the onSignUp', () => {
      expect(mockSignUp).toBeCalledTimes(1);
    });
  });
});
