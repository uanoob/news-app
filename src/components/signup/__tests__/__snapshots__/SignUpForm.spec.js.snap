// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp render correctly SignUpFormComponent component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUpFormComponent
    classes={
      Object {
        "container": "",
        "textField": "",
      }
    }
    onSignUp={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "autoComplete": "off",
      "children": Array [
        <TextField
          autoComplete="current-name"
          className=""
          error={false}
          id="outlined-name-input"
          label="Name"
          margin="normal"
          onChange={[Function]}
          required={false}
          select={false}
          type="text"
          value=""
          variant="outlined"
        />,
        <TextField
          autoComplete="email"
          className=""
          error={false}
          id="outlined-email-input"
          label="Email"
          margin="normal"
          name="email"
          onChange={[Function]}
          required={false}
          select={false}
          type="email"
          value=""
          variant="outlined"
        />,
        <TextField
          autoComplete="current-password"
          className=""
          error={false}
          id="outlined-password-input"
          label="Password"
          margin="normal"
          onChange={[Function]}
          required={false}
          select={false}
          type="password"
          value=""
          variant="outlined"
        />,
        <WithStyles(Button)
          color="primary"
          disabled={true}
          id="signup-button"
          type="submit"
          variant="contained"
        >
          Sign Up
        </WithStyles(Button)>,
      ],
      "className": "",
      "noValidate": true,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoComplete": "current-name",
          "className": "",
          "error": false,
          "id": "outlined-name-input",
          "label": "Name",
          "margin": "normal",
          "onChange": [Function],
          "required": false,
          "select": false,
          "type": "text",
          "value": "",
          "variant": "outlined",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoComplete": "email",
          "className": "",
          "error": false,
          "id": "outlined-email-input",
          "label": "Email",
          "margin": "normal",
          "name": "email",
          "onChange": [Function],
          "required": false,
          "select": false,
          "type": "email",
          "value": "",
          "variant": "outlined",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoComplete": "current-password",
          "className": "",
          "error": false,
          "id": "outlined-password-input",
          "label": "Password",
          "margin": "normal",
          "onChange": [Function],
          "required": false,
          "select": false,
          "type": "password",
          "value": "",
          "variant": "outlined",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Sign Up",
          "className": undefined,
          "color": "primary",
          "disabled": true,
          "id": "signup-button",
          "type": "submit",
          "variant": "contained",
        },
        "ref": null,
        "rendered": "Sign Up",
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "autoComplete": "off",
        "children": Array [
          <TextField
            autoComplete="current-name"
            className=""
            error={false}
            id="outlined-name-input"
            label="Name"
            margin="normal"
            onChange={[Function]}
            required={false}
            select={false}
            type="text"
            value=""
            variant="outlined"
          />,
          <TextField
            autoComplete="email"
            className=""
            error={false}
            id="outlined-email-input"
            label="Email"
            margin="normal"
            name="email"
            onChange={[Function]}
            required={false}
            select={false}
            type="email"
            value=""
            variant="outlined"
          />,
          <TextField
            autoComplete="current-password"
            className=""
            error={false}
            id="outlined-password-input"
            label="Password"
            margin="normal"
            onChange={[Function]}
            required={false}
            select={false}
            type="password"
            value=""
            variant="outlined"
          />,
          <WithStyles(Button)
            color="primary"
            disabled={true}
            id="signup-button"
            type="submit"
            variant="contained"
          >
            Sign Up
          </WithStyles(Button)>,
        ],
        "className": "",
        "noValidate": true,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": "current-name",
            "className": "",
            "error": false,
            "id": "outlined-name-input",
            "label": "Name",
            "margin": "normal",
            "onChange": [Function],
            "required": false,
            "select": false,
            "type": "text",
            "value": "",
            "variant": "outlined",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": "email",
            "className": "",
            "error": false,
            "id": "outlined-email-input",
            "label": "Email",
            "margin": "normal",
            "name": "email",
            "onChange": [Function],
            "required": false,
            "select": false,
            "type": "email",
            "value": "",
            "variant": "outlined",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": "current-password",
            "className": "",
            "error": false,
            "id": "outlined-password-input",
            "label": "Password",
            "margin": "normal",
            "onChange": [Function],
            "required": false,
            "select": false,
            "type": "password",
            "value": "",
            "variant": "outlined",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Sign Up",
            "className": undefined,
            "color": "primary",
            "disabled": true,
            "id": "signup-button",
            "type": "submit",
            "variant": "contained",
          },
          "ref": null,
          "rendered": "Sign Up",
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
